### PR TITLE
Enable multiple inputs in AstQuery

### PR DESCRIPTION
**Pull request by lukedirtwalker**
This makes it possible to put multiple inputs to an ast query, and the
query is applied to all input.

I decided to a separate PR for this since it got a bit larger than intended.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40418901%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40419015%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40419077%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23issuecomment-143188498%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23issuecomment-143189142%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23issuecomment-143189491%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23issuecomment-143189611%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%2004badb8ca0949e380feac37a176c1dfb028ab3cd%20InformationScripting/src/queries/AstQuery.cpp%20122%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40419015%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22error/warning%22%2C%20%22created_at%22%3A%20%222015-09-25T10%3A58%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL193-325%22%7D%2C%20%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23issuecomment-143188498%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hmm%2C%20now%20that%20I%20see%20this%2C%20I%27m%20wondering%20if%20we%27re%20doing%20the%20right%20thing%20here.%20For%20many%20%28most%3F%29%20commands%2C%20it%27s%20clear%20that%20having%20more%20than%20one%20input%20is%20either%3A%5Cr%5Cn-%20completely%20unecessary%5Cr%5Cn-%20just%20doing%20the%20same%20thing%20multiple%20times%2C%20independently%20of%20each%20other.%5Cr%5Cn%5Cr%5CnMultiple%20inputs%20for%20a%20command%20are%20only%20interesting%2C%20when%20the%20different%20inputs%20interact%2C%20or%20somehow%20have%20different%20meanings.%20Otherwise%20we%20can%20just%20run%20the%20same%20command%20multiple%20times%20to%20achieve%20the%20same%20effect%20%2B%20we%20can%20write%20the%20commands%20in%20a%20simpler%20way.%5Cr%5Cn%5Cr%5CnSo%20I%27m%20thinking%2C%20instead%20of%20handling%20multiple%20inputs%20in%20these%20commands%2C%20where%20I%20don%27t%20think%20we%20gain%20something%2C%20we%20should%20make%20it%20more%20explicit%20that%20they%20only%20support%20a%20single%20input%2C%20and%20call%20them%20multiple%20times%20if%20neccessary.%5Cr%5Cn%5Cr%5CnWe%20could%20enhance%20the%20interface%20of%20%60Query%60%20to%20have%20methods%20that%20talk%20about%20the%20single%20%60input%60/%60output%60%20and%20just%20connect%20this%20to%20inputs%5B0%5D%20and%20outputs%5B0%5D%20behind%20the%20scenes.%5Cr%5Cn%5Cr%5CnAlternatively%20%28and%20probably%20better%29%20we%20could%20introduce%20a%20class%20%60LinearQuery%20%3A%20Query%60%20that%20%20wraps%20the%20existing%20multiple%20input/output%20interface%20and%20exposes%20a%20new%20simpler%20interface%20for%20single%20in/single%20out%2C%20and%20use%20that%20for%20most%20commands.%20This%20way%2C%20we%20could%20even%20centralize%20checks%20that%20violate%20our%20assumptions%20of%20sinlge%20input/output.%20Or%20if%20we%27re%20smart%2C%20we%20could%20just%20automatically%20perform%20the%20computation%20multple%20times.%20Either%20way%2C%20the%20%5C%22client%5C%22%20will%20only%20have%20to%20work%20with%20a%20simplified%20interface%2C%20which%20makes%20sense.%20So%20in%20essense%20this%20will%20abstract/remove%20the%20many%20identical%20loops%20that%20you%20are%20adding%20with%20this%20PR.%5Cr%5Cn%5Cr%5CnWhat%20do%20you%20think%3F%22%2C%20%22created_at%22%3A%20%222015-09-25T11%3A08%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes%20I%20like%20this%20idea%20about%20the%20LinearQuery.%20I%20will%20see%20how%20it%20works%2C%20once%20I%20finished%20the%20tags%20stuff.%22%2C%20%22created_at%22%3A%20%222015-09-25T11%3A12%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22OK%2C%20In%20that%20case%20should%20we%20%5C%22postpone%5C%22%20the%20current%20PR%20until%20then%2C%20or%20would%20you%20prefer%20to%20merge%20this%20and%20change%20it%20later%3F%22%2C%20%22created_at%22%3A%20%222015-09-25T11%3A15%3A42Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22No%20we%20don%27t%20want%20to%20merge%20this.%22%2C%20%22created_at%22%3A%20%222015-09-25T11%3A16%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%2004badb8ca0949e380feac37a176c1dfb028ab3cd%20InformationScripting/src/queries/AstQuery.cpp%2081%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40418901%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20guess%20a%20warning%20is%20better.%22%2C%20%22created_at%22%3A%20%222015-09-25T10%3A57%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL193-325%22%7D%2C%20%22Pull%2004badb8ca0949e380feac37a176c1dfb028ab3cd%20InformationScripting/src/queries/AstQuery.cpp%20177%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/156%23discussion_r40419077%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22error/warning%22%2C%20%22created_at%22%3A%20%222015-09-25T10%3A59%3A24Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/AstQuery.cpp%3AL193-325%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-Pull 04badb8ca0949e380feac37a176c1dfb028ab3cd InformationScripting/src/queries/AstQuery.cpp 81'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/156#discussion_r40418901'>File: InformationScripting/src/queries/AstQuery.cpp:L193-325</a></b>
- <a href='https://github.com/dimitar-asenov'></a> I guess a warning is better.
- [ ] <a href='#crh-comment-Pull 04badb8ca0949e380feac37a176c1dfb028ab3cd InformationScripting/src/queries/AstQuery.cpp 122'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/156#discussion_r40419015'>File: InformationScripting/src/queries/AstQuery.cpp:L193-325</a></b>
- <a href='https://github.com/dimitar-asenov'></a> error/warning
- [ ] <a href='#crh-comment-Pull 04badb8ca0949e380feac37a176c1dfb028ab3cd InformationScripting/src/queries/AstQuery.cpp 177'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/156#discussion_r40419077'>File: InformationScripting/src/queries/AstQuery.cpp:L193-325</a></b>
- <a href='https://github.com/dimitar-asenov'></a> error/warning
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/156#issuecomment-143188498'>General Comment</a></b>
- <a href='https://github.com/dimitar-asenov'></a> Hmm, now that I see this, I'm wondering if we're doing the right thing here. For many (most?) commands, it's clear that having more than one input is either:
- completely unecessary
- just doing the same thing multiple times, independently of each other.
  Multiple inputs for a command are only interesting, when the different inputs interact, or somehow have different meanings. Otherwise we can just run the same command multiple times to achieve the same effect + we can write the commands in a simpler way.
  So I'm thinking, instead of handling multiple inputs in these commands, where I don't think we gain something, we should make it more explicit that they only support a single input, and call them multiple times if neccessary.
  We could enhance the interface of `Query` to have methods that talk about the single `input`/`output` and just connect this to inputs[0] and outputs[0] behind the scenes.
  Alternatively (and probably better) we could introduce a class `LinearQuery : Query` that  wraps the existing multiple input/output interface and exposes a new simpler interface for single in/single out, and use that for most commands. This way, we could even centralize checks that violate our assumptions of sinlge input/output. Or if we're smart, we could just automatically perform the computation multple times. Either way, the "client" will only have to work with a simplified interface, which makes sense. So in essense this will abstract/remove the many identical loops that you are adding with this PR.
  What do you think?
- <a href='https://github.com/lukedirtwalker'></a> Yes I like this idea about the LinearQuery. I will see how it works, once I finished the tags stuff.
- <a href='https://github.com/dimitar-asenov'></a> OK, In that case should we "postpone" the current PR until then, or would you prefer to merge this and change it later?
- <a href='https://github.com/lukedirtwalker'></a> No we don't want to merge this.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/156?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/156?approve=1'></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/156'></a>
<a href='#crh-end'></a>

### CODE REVIEW

File: InformationScripting/src/queries/AstQuery.cpp [hunk 1]
```diff
@@ -191,101 +193,132 @@ QList<TupleSet> AstQuery::genericQuery(QList<TupleSet> input)
...
+	}
 	else if (scope() == Scope::Input)
 	{
-		Q_ASSERT(input.size());
-		tuples = input.takeFirst();
-
-		// Note here we remove the input nodes.
-		auto tuple = tuples.take("ast");
-		for (const auto& t : tuple) addNodesOfType(tuples, matcher, t["ast"]);
+		Q_ASSERT(input.size()); // TODO: user warning or ignore?
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/AstQuery.cpp:213]: I guess a warning is better.

File: InformationScripting/src/queries/AstQuery.cpp [hunk 2]
```diff
@@ -191,101 +193,132 @@ QList<TupleSet> AstQuery::genericQuery(QList<TupleSet> input)
...
+	}
 	else if (scope() == Scope::Input)
 	{
-		Q_ASSERT(input.size());
-		tuples = input.takeFirst();
-
-		// Note here we remove the input nodes.
-		auto tuple = tuples.take("ast");
-		for (const auto& t : tuple) matchingNodes << Model::NameResolver::mostLikelyMatches(name, -1, t["ast"]);
+		Q_ASSERT(input.size()); // TODO warn user or ignore?
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/AstQuery.cpp:242]: error/warning

File: InformationScripting/src/queries/AstQuery.cpp [hunk 3]
```diff
@@ -191,101 +193,132 @@ QList<TupleSet> AstQuery::genericQuery(QList<TupleSet> input)
...
 	{
-		Q_ASSERT(input.size());
-		auto tuples = input.takeFirst();
-
-		auto tuple = tuples.tuples("ast");
-		for (const auto& t : tuple) references[t["ast"]] = AllNodesOfType::allNodesOfType<Model::Reference>(t["ast"]);
-
-		// Keep all non ast nodes
-		result = tuples;
+		Q_ASSERT(input.size()); // TODO warn user or ignore ?
```
Comment (dimitar-asenov) [inline InformationScripting/src/queries/AstQuery.cpp:280]: error/warning

**Comment by dimitar-asenov** [comment]
Hmm, now that I see this, I'm wondering if we're doing the right thing here. For many (most?) commands, it's clear that having more than one input is either:
- completely unecessary
- just doing the same thing multiple times, independently of each other.

Multiple inputs for a command are only interesting, when the different inputs interact, or somehow have different meanings. Otherwise we can just run the same command multiple times to achieve the same effect + we can write the commands in a simpler way.

So I'm thinking, instead of handling multiple inputs in these commands, where I don't think we gain something, we should make it more explicit that they only support a single input, and call them multiple times if neccessary.

We could enhance the interface of `Query` to have methods that talk about the single `input`/`output` and just connect this to inputs[0] and outputs[0] behind the scenes.

Alternatively (and probably better) we could introduce a class `LinearQuery : Query` that  wraps the existing multiple input/output interface and exposes a new simpler interface for single in/single out, and use that for most commands. This way, we could even centralize checks that violate our assumptions of sinlge input/output. Or if we're smart, we could just automatically perform the computation multple times. Either way, the "client" will only have to work with a simplified interface, which makes sense. So in essense this will abstract/remove the many identical loops that you are adding with this PR.

What do you think?

**Comment by lukedirtwalker** [comment]
Yes I like this idea about the LinearQuery. I will see how it works, once I finished the tags stuff.

**Comment by dimitar-asenov** [comment]
OK, In that case should we "postpone" the current PR until then, or would you prefer to merge this and change it later?

**Comment by lukedirtwalker** [comment]
No we don't want to merge this.

**Pull request closed without merging**